### ISSUE TITLE
Count can be wrong in Zend Config

### ISSUE THREAD

**Issue by Logiqu** [zendframework/zf2#7253]
When setting the same key on a config object the counter always increments without caring about already existing keys:

```php
use Zend\Config\Config;

$config = new Config([], true);

$config->foo = 1;
$config->foo = 2;
assert($config->foo == 2);
assert(count($config->toArray()) == 1);
assert(count($config) == 1);
```

The issue can be found in `__set` of class `Zend\Config\Config`.
There are two possible solutions for this issue:

1. decrease the counter if the key is already defined:

    ```php
    if (null === $name) {
        $this->data[] = $value;
    } else {
        if (array_key_exists($name, $this->data)) {
            $this->count--;
        }
        $this->data[$name] = $value;
    }

    $this->count++;
```

2. remove the counter and call `count($this->data)` in the `count()` method of the class

**Comment by fabiang** [comment]
This issue did not exists in Zend Framework 1. [`count()` was always called on `$this->_data` at every single call of `__set`](https://github.com/zendframework/zf1/blob/master/library/Zend/Config.php#L165)

Is the better performance the reason why `count()` is not simply called on the data:
```php
public function count()
{
    return count($this->data);
}
```
?

**Comment by weierophinney** [comment]
Fixed with #7256
